core/src/main/scala/kafka/log/Log.scala

```diff
@@ -609,24 +716,22 @@ class Log(val dir: File,
    * logSegment will be rolled if one of the following conditions met
    * <ol>
    * <li> The logSegment is full
-   * <li> The maxTime has elapsed
+   * <li> The maxTime has elapsed since the largest timestamp in the segment (or since the create time if there is no
+   * message having a timestamp)
    * <li> The index is full
    * </ol>
    * @return The currently active segment after (perhaps) rolling to a new segment
    */
   private def maybeRoll(messagesSize: Int): LogSegment = {
     val segment = activeSegment
+    val reachedRollMs = segment.timeWaitedForRoll(time.milliseconds) > config.segmentMs - segment.rollJitterMs
```

Comment:
Yes, what you suggested sounds like a good idea. I would just modify that slightly to do the time-based rolling based on the timestamp of the first message in the segment. We probably want to update the KIP wiki and bring this up in the mailing list to see if people have any concern about the change.